colossalai/fx/passes/algorithms/linearize.py

```diff
@@ -0,0 +1,89 @@
+    for node in gm.graph.nodes:
+        last_dict_len = len(status_dict)
+        # remove node from users list in status_dict
+        for item in status_dict.values():
+            if node in item:
+                item.remove(node)
+
+        # pop node from status_dict if it is fully used
+        for key in list(status_dict):
+            if len(status_dict[key]) == 0:
+                status_dict.pop(key)
+
+        # first node in graph, it should be in n0-n1 type,
+        # where n0 contains only input op, i.e. placeholder
+        if last_dict_len == 0:
+            node_dict[node_idx] = [node]
+            status_dict[node.name] = list(node.users)
+            node_idx += 1
+            node_dict[node_idx] = []
+
+            continue
+
+        # boundary case
+        if len(status_dict) == 0:
+            # current node region end point = next node region start point
+            # i.e. n1-n2-n3-... type node, each node contains only one op
+            if last_dict_len == 1:
+                if len(node_dict[node_idx]) > 0:
+                    node_idx += 1
+                    node_dict[node_idx] = []
+                node_dict[node_idx].append(node)
+                status_dict[node.name] = list(node.users)
+
+                continue
+
+            # n1-n2-n3, if n1 has multiple ops, the last op in n1 will be
+            # the one who is able to clean all others in status_dict
+            # and as the last_dict_len > 1, there are multiple ops are used
+            # by this node, we view it as the end of one node and start a new node
+            else:
+
+                node_dict[node_idx].append(node)
+                status_dict[node.name] = list(node.users)
+                node_idx += 1
+                node_dict[node_idx] = []
+
+                continue
+
+        else:
+            # currently I will use bigger node structure
+            # if the following region is activated, the node will be smaller
+            #################################################
+            # if last_dict_len == 1:
+            #     if len(node_dict[node_idx]) > 0:
+            #         node_idx += 1
+            #     node_dict[node_idx] = [node]
+            #     status_dict[node.name] = list(node.users)
+            #
+            #     continue
+            #################################################
+
+            # in-node case, as the current node can not clean status_dict
+            # we view it as in-node status, the node will be appended to the
+            # current node_idx
+            node_dict[node_idx].append(node)
+            status_dict[node.name] = list(node.users)
+
+            continue
+
+    # If the output node use multiple nodes, there might be an
+    # empty node after the output node
+    if len(node_dict[node_idx]) == 0:
+        node_dict.pop[node_idx]
+        node_idx -= 1
+
+    # pop the last two nodes
+    node_dict.pop(0)
+    node_dict.pop(node_idx)
+    return node_dict
```

Comment:
I believe the following code works the same.
``` python
def linearize(gm: GraphModule) -> List[List[Node, ...]]:
    """
    In most existing frameworks of activation checkpoint, the forward graph is assumed to be linearized.
    """

    def is_sink():
        """
        If we can free all memories when executing a certain node, it is a sink.
        """
        return not sum((v for k, v in deps.items()))

    deps = {}
    linearized_nodes = []
    region = []
    for n in gm.graph.nodes:
        for n_par in n._input_nodes:
            deps[n_par] -= 1    # free memory and dependencies
        region.append(n)    # append the node to this region

        # we can call an end to the region if the current node is the only living node in memory without any dependencies
        if is_sink():
            linearized_nodes.append(region)
            region = []
        deps[n] = len(n.users)    # add dependencies for future executions
    return linearized_nodes
```
The `linearized_nodes: List[List[Node, ...]]` will contain regions of nodes like `[[node_0, node_1, node_2], [node_3, node_4], ...]`.

Reply:
Great idea, but currently my code maintains some flexibility if we need smaller structure of node, check the annotated region line 62-70, in your code above, the below case will be considered as two nodes:
              -n3-
n0 - n1 - n2 - n4
but if we activate line 62-70, n1 could be seen as a node too, I think currently I will maintain my linearize structure, as the `solver_rotor` was built with the `node_dict`, I will consider your simplified code in next PR.

Reply:
Great idea, I think it is much better than my code, but currently I deactivate line 62-70, so there will be less nodes in our linearized graph, which could might be better for our solver right now. I will consider your simplified structure later.